File: web/domain/ocrAmountLearner.js
import { roundToCents } from './moneyMath.js';
import { copyOrDefault, clampWeights, FEATURE_COUNT, STEP } from './ocrAmountWeights.js';
import { listMoneyCandidates } from './receiptFieldParser.js';

/**
 * Updates OCR fallback weights when the saved amount differs from the OCR guess
 * and both amounts appear as money candidates on the lines.
 */
export function learn(lines, ocrAmount, savedAmount, currentWeights, mode) {
  const current = copyOrDefault(currentWeights);
  if (ocrAmount == null || !lines || lines.length === 0) return current;
  if (amountsMatch(ocrAmount, savedAmount)) return current;
  const candidates = listMoneyCandidates(lines, mode);
  const rejected = findByAmount(candidates, ocrAmount);
  const chosen = findByAmount(candidates, savedAmount);
  if (!rejected || !chosen) return current;
  const chosenFlags = featureFlags(chosen);
  const rejectedFlags = featureFlags(rejected);
  const next = copyOrDefault(current);
  for (let i = 0; i < FEATURE_COUNT; i++) {
    if (chosenFlags[i] && !rejectedFlags[i]) next[i] += STEP;
    else if (rejectedFlags[i] && !chosenFlags[i]) next[i] -= STEP;
  }
  return clampWeights(next);
}

function amountsMatch(left, right) {
  return Math.abs(roundToCents(left) - roundToCents(right)) < 0.001;
}

function findByAmount(candidates, amount) {
  return candidates.find((c) => amountsMatch(c.amount, amount)) || null;
}

function featureFlags(candidate) {
  return [
    candidate.dollarSign,
    candidate.strongTotalLabel,
    candidate.totalLabel,
    candidate.bottomHalf,
    candidate.orderOrPoints,
  ];
}

File: app/src/main/java/com/example/envelopemoney/receipt/OcrAmountLearner.java
package com.example.envelopemoney.receipt;

import com.example.envelopemoney.MoneyMath;

import java.util.List;

/**
 * Updates OCR fallback weights when the user saves a different amount than OCR guessed.
 *
 * <p>No change if the saved amount is not a money candidate on the lines, if OCR never ran,
 * or if the two amounts match at cents. Side-effect free besides the returned vector.
 */
public final class OcrAmountLearner {

    private OcrAmountLearner() {
    }

    public static float[] learn(List<String> lines,
                                Double ocrAmount,
                                double savedAmount,
                                float[] currentWeights,
                                ReceiptCaptureMode mode) {
        float[] current = OcrAmountWeights.copyOrDefault(currentWeights);
        if (ocrAmount == null || lines == null || lines.isEmpty()) {
            return current;
        }
        if (amountsMatch(ocrAmount, savedAmount)) {
            return current;
        }
        List<OcrMoneyCandidate> candidates = ReceiptFieldParser.listMoneyCandidates(lines, mode);
        OcrMoneyCandidate rejected = findByAmount(candidates, ocrAmount);
        OcrMoneyCandidate chosen = findByAmount(candidates, savedAmount);
        if (rejected == null || chosen == null) {
            return current;
        }
        boolean[] chosenFlags = chosen.featureFlags();
        boolean[] rejectedFlags = rejected.featureFlags();
        float[] next = OcrAmountWeights.copyOrDefault(current);
        for (int i = 0; i < OcrAmountWeights.FEATURE_COUNT; i++) {
            if (chosenFlags[i] && !rejectedFlags[i]) {
                next[i] += OcrAmountWeights.STEP;
            } else if (rejectedFlags[i] && !chosenFlags[i]) {
                next[i] -= OcrAmountWeights.STEP;
            }
        }
        return OcrAmountWeights.clamp(next);
    }

    private static boolean amountsMatch(double left, double right) {
        return Math.abs(MoneyMath.roundToCents(left) - MoneyMath.roundToCents(right)) < 0.001d;
    }

    private static OcrMoneyCandidate findByAmount(List<OcrMoneyCandidate> candidates, double amount) {
        for (OcrMoneyCandidate candidate : candidates) {
            if (amountsMatch(candidate.amount, amount)) {
                return candidate;
            }
        }
        return null;
    }
}
